File: Limbus/limbus_core/frontend/Token.fs
module Token
open Source

type TokenType = int
type TokenValue = I of int | F of float | None

let tchar2s(t : TCHAR) : string =
    match t with 
        | EOL | EOF -> ""
        | _ -> string(C)    

[<AbstractClass>]
type Token(source : Source) = 
    class
        let mutable _type : TokenType = 0
        let mutable _text : string = ""
        let mutable _value : TokenValue = I 0
        let mutable _source = source
        let mutable _line_num = _source.line_num
        let mutable _position = _source.current_pos
        
        member this.extract()  = 
            _text <- this.currnet_char() |> tchar2s
            _value <- None
            this.next_char()
                    
        member this.currnet_char() : TCHAR =
            _source.current_char()
    
        member this.next_char() : TCHAR  =
            _source.next_char()
            
        member this.peek_char() : TCHAR =
            _source.peek_char()
    end

[<Class>]
type EodOfToken(source : Source) =
    inherit Token(source)
    

    member this.extract() = 
        base.extract()

File: Limbus/limbus_core/frontend/token.py
# -*- coding: utf-8 -*-
from enum import Enum,  auto

class TokenType(Enum):
    DUMMY = auto()
    EOF = auto()
    ERROR = auto()
    PASCAL = auto()


class Token:
    def __init__(self, source):
        self.type = TokenType.DUMMY
        self.text = ""
        self.value = None
        self.error_code = None
        self.source = source
        self.line_num = source.get_line_num()
        self.pos = source.get_position()
        self.extract()

    def extract(self):
        self.text = self.current_char()
        self.value = 0
        self.next_char()

    def current_char(self):
        return self.source.current_char()

    def next_char(self):
        return self.source.next_char()

    def peek_char(self):
        return self.source.peek_char()


class ErrorToken(Token):
    def __init__(self, err_code, source):
        super().__init__(source)
        self.type = TokenType.ERROR
        self.err_code = err_code


class EofToken(Token):
    def __init__(self, source):
        super().__init__(source)
        self.type = TokenType.EOF


